### PR TITLE
feat(ui/data-tabe): Add initial filtering to datasets

### LINKED ISSUE #194: Filter datasets
# Add the ability to filter datasets by a range of properties

Currently all datasets that are visible to a user are displayed. This will become a very large number so we must add the ability to filter these by a range or properties. These filters will appear above the table that displays the datasets.

The current text filter box will remain and will filter only those datasets that are displayed in the same way that if currently does. It should be distinct from the filter elements as this behaviour is different.

Also related to this is the ability to paginate the results, but this is not yet supported by the API (1) 

## Initial properties to filter
Filtering by these properties is already supported by the API.
- Owner (select one user from the list of users) (2)
- Mime-type (Note: we should use human friendly names here not terms like chemical/x-mdl-sdfile). This will be a drop down list with values coming from the `/type` endpoint
- Label name (list of comma separated labels). It would be nice to have some sort of autocomplete here, but the full list of labels is not currently available (3)

## Filter properties that will be added later
These filters should soon become available
- Editor (similar to how the owner filter works)
- Label name=value

## A note about labels
Labels are simple name=value pairs with the value being optional. Hence these are valid examples:
`important`
`pdb`
`pdb=1HC3`

A dataset can have any number of labels.

Currently only filtering by the label name is supported by the API, but we will soon be adding the ability to specify a name and a value. e.g. if you specify `pdb` as a label name filter you would get the last two items, but if you specified `pdb=1HC3` then you would only get the last item.

## IM tasks

1. @duncanpeacock update this (maybe create new issue?) once pagination is supported
2. @OliverDudgeon describe how to get the current list of users
3. @duncanpeacock examine if a list of label names can be made accessible

### PR DESCRIPTION
**Pull request by ag-m2ms**
Resolves #194

### CODE REVIEW

File: components/DatasetsTable/DatasetsToolbar.tsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+import { css } from '@emotion/react';
+import { Grid, useTheme } from '@material-ui/core';
+
+export interface DatasetsToolbarProps {
+  /**
+   * Elements to be wrapped by the component.
+   */
+  children: ReactNode;
+}
```
Comment (OliverDudgeon) [inline components/DatasetsTable/DatasetsToolbar.tsx:11]: For consistency use FC from react when using children.

File: components/DatasetsTable/filters/LabelFilter.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * Component which adjusts filtering of datasets according to label.
+ */
+export const LabelFilter = ({ label, setLabel }: LabelFilterProps) => {
+  // Since we are using debounce, value in the 'label' does not update in real time.
+  // That is why we use this state variable which displays the provided text in real time.
+  const [displayLabel, setDisplayLabel] = useState(label);
+
+  const onChangeDebounced = useMemo(() => {
```
Comment (OliverDudgeon) [inline components/DatasetsTable/filters/LabelFilter.tsx:27]: Nothing really "wrong" here but I usually only use memo/callback hooks when fixing a performance issue or render bug. If that's not what you're trying to do with it here then it seems unnecessary. If so disgegard this.

File: components/DatasetsTable/filters/index.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+export * from './FileTypeFilter';
+export * from './LabelFilter';
```
Comment (OliverDudgeon) [inline components/DatasetsTable/filters/index.ts:2]: Just for convention, I've typically used an index file when a folder corresponds to *one* main component and I then re-export that in the index.

File: components/DatasetsTable/DatasetsToolbar.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * Wraps elements in the DataTable toolbar and adds spacing between them.
+ */
+export const DatasetsToolbar = ({ children }: DatasetsToolbarProps) => {
+  const theme = useTheme();
+  return (
+    <Grid
+      container
+      alignItems="center"
+      css={css`
+        gap: ${theme.spacing()}px;
```
Comment (OliverDudgeon) [inline components/DatasetsTable/DatasetsToolbar.tsx:23]: Any reason to not use the spacing property of a Grid here? Also I think "row" is the default direction.

File: components/DatasetsTable/filters/index.ts (hunk 3 above)
Comment (ag-m2ms) [reply components/DatasetsTable/filters/index.ts:2]: So basically I just remove the index file and reimport the components? Other than that, is the file structure corresponding to the rest of the project?

File: components/DatasetsTable/DatasetsToolbar.tsx (hunk 4 above)
Comment (OliverDudgeon) [reply components/DatasetsTable/DatasetsToolbar.tsx:23]: I think the usage of Grid can be confusing here without the Grid items. I'd switch to wrapping grid items with spacing prop. If you want to reduce the complexity in the parent scope, you could map over the passed children prop and wrap each child in a grid item.

**Pull request merged**